pkg/console/controllers/route/controller.go

```diff
@@ -110,59 +122,151 @@ func (c *RouteSyncController) sync() error {
+	return defaultRoute, defaultRouteChanged, "", defaultRouteError
+}
+
+// Custom route sync needs to:
+// 1. validate if the reference for secret with serving-cert is defined in operator config(in case a non-openshift cluster domain is used)
+// 2. if secret is defined, validate if the secret contains TLS certificate and key
+// 3. if secret is defined, get the TLS certificate and key by bese64 decoding them
+// 4. create the custom console route, if custom TLS certificate and key are defined use them
+// 5. validate and stomp the route
+func (c *RouteSyncController) SyncCustomRoute(operatorConfig *operatorsv1.Console) (consoleRoute *routev1.Route, isNew bool, reason string, err error) {
+	if !routesub.IsCustomRouteSet(operatorConfig) {
+		return nil, false, "", nil
+	}
+
+	customSecret, configErr := c.ValidateCustomRouteConfig(operatorConfig)
+	if configErr != nil {
+		return nil, true, "InvalidCustomRouteConfig", configErr
+	}
+
+	customTLSCert, secretValidationErr := c.ValidateCustomCertSecret(customSecret)
+	if secretValidationErr != nil {
+		return nil, true, "InvalidCustomCertificateSecret", secretValidationErr
+	}
+
+	requiredCustomRoute := routesub.CustomRoute(operatorConfig, customTLSCert)
+
+	customRoute, customRouteChanged, customRouteError := routesub.ApplyRoute(c.routeClient, c.recorder, requiredCustomRoute)
+	if customRouteError != nil {
+		return nil, false, "FailedApply", customRouteError
 	}
 
-	if validatedRoute, changed := routesub.Validate(rt); changed {
-		// if validation changed the route, issue an update
-		if _, err := c.routeClient.Routes(api.TargetNamespace).Update(c.ctx, validatedRoute, metav1.UpdateOptions{}); err != nil {
-			// error is unexpected, this is a real error
-			return nil, false, "InvalidRouteCorrection", err
+	return customRoute, customRouteChanged, "", customRouteError
+}
+
+func (c *RouteSyncController) ValidateCustomRouteConfig(operatorConfig *operatorsv1.Console) (*corev1.Secret, error) {
+	// get ingress
+	ingress, err := c.ingressClient.Get(c.ctx, api.ConfigResourceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Check if the custom hostname has cluster domain suffix. If it does,
+	// then the secret that contains TLS certificate and key needs to be
+	// specified on the operator config.
+	if strings.HasSuffix(operatorConfig.Spec.Route.Hostname, ingress.Spec.Domain) {
+		return nil, nil
+	}
+	if !routesub.IsCustomRouteSecretSet(operatorConfig) {
+		return nil, fmt.Errorf("secret reference for custom route certificate is not defined")
+	}
+	customCertSecret, err := c.secretClient.Secrets(api.OpenShiftConfigNamespace).Get(c.ctx, operatorConfig.Spec.Route.Secret.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return customCertSecret, nil
+}
+
+// Validate secret that holds custom TLS certificate and key.
+// Secret has to contain `tls.crt` and `tls.key` data keys
+// where the certificate and key are stored.
+// Return the custom TLS certificate and key
```

Comment:
Nice!  Tested this with a dummy bad file?

Reply:
about to add !